rapids-cmake/test/add.cmake

```diff
@@ -0,0 +1,149 @@
+#=============================================================================
+# Copyright (c) 2022-2023, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#=============================================================================
+include_guard(GLOBAL)
+
+#[=======================================================================[.rst:
+rapids_test_add
+---------------
+
+.. versionadded:: v23.04.00
+
+States how many GPUs and what percent of each a test requires.
+
+  .. code-block:: cmake
+
+    rapids_test_add(NAME <name> COMMAND <target|command> [<args>...]
+                    [GPUS <N> [PERCENT <value>]]
+                    [INSTALL_COMPONENT_SET <set>]
+                    [WORKING_DIRECTORY <dir>])
+
+Add a test called `<name>` which will be executed with a given GPU
+resource allocation.
+
+When combined with :cmake:command:`rapids_test_init` informs CTest what
+resources should be allocated to a test so that when testing in parallel
+oversubscription doesn't occur. Without this information user execution of
+CTest with high parallel levels will cause multiple tests to run on the
+same GPU and quickly exhaust all memory.
+
+``COMMAND``
+  Specify the test command-line. If <command> specifies an executable target
+  (created by add_executable()) it will automatically be replaced by the location of the
+  executable created at build time.
+
+``GPUS``
+  State how many GPUs this test requires. Allows CTest to not over-subscribe
+  a machine's hardware.
+
+  If no value is provided, the test is considered to not use any GPUs
+
+``PERCENT``
+  By default if no percent is provided, 100 is used.
+  State how much of each GPU this test requires.
```

Comment:
I was initially thinking that it would make for a nicer API if we could use a floating point value for GPUs to indicate fractional usage, but seeing how you've implemented the translation to a resource group with 100 slots per GPU I see why you went with percent as a separate integral argument. Do you think it would introduce too much internal complexity to be worth implementing that API? I can see the complexities around needing to round and validate a float.

Reply:
One advantage of the float approach is that it removes some ambiguity over what argument sets are valid. For instance, does it make sense to do `GPUS 2 PERCENT 25` as compared to `GPUS 1 PERCENT 50`? This may also be construed as a disadvantage if you intended to allow that kind of spec where a test requires multiple fractional GPUs, though.

Reply:
I prefer the percentage approach as it keeps the CMake logic simple. Mapping from decimals back to integer slots will cause edge cases around rounding and unreprenstable value ( e.g. 0.1  not mapping to a max of 10 jobs concurrently).